kie-soup-dataset/kie-soup-dataset-prometheus/src/main/java/org/dashbuilder/dataprovider/prometheus/PrometheusDataSetProvider.java

```diff
@@ -0,0 +1,165 @@
+            SINGLETON = new PrometheusDataSetProvider(staticDataSetProvider);
+            dataSetDefRegistry.addListener(SINGLETON);
+        }
+        return SINGLETON;
+    }
+
+    public PrometheusDataSetProvider() {}
+
+    public PrometheusDataSetProvider(StaticDataSetProvider staticDataSetProvider) {
+        this.staticDataSetProvider = staticDataSetProvider;
+    }
+
+    @SuppressWarnings("rawtypes")
+    public DataSetProviderType getType() {
+        return DataSetProviderType.PROMETHEUS;
+    }
+
+    public DataSetMetadata getDataSetMetadata(DataSetDef def) throws Exception {
+        DataSet dataSet = lookupDataSet(def, null);
+        if (dataSet == null) {
+            return null;
+        }
+        return dataSet.getMetadata();
+    }
+
+    public DataSet lookupDataSet(DataSetDef def, DataSetLookup lookup) throws Exception {
+
+        // TODO: Implement cache - if cache or test then return the dataset def from static 
+        // - otherwise go to the server and get the dataset
+        // Look first into the static data set provider since CSV data set are statically registered once loaded.
+        String baseUrl = ((PrometheusDataSetDef) def).getServerUrl();
+        String query = ((PrometheusDataSetDef) def).getQuery();
+        QueryResponse response = new PrometheusClient(baseUrl).query(query);
+
+        if (response.getStatus() == Status.ERROR) {
+            throw new IllegalArgumentException("Error response received from Prometheus: " + response.getError());
+        }
+
+        DataSet dataSet = toDataSet(response);
+        dataSet.setUUID(def.getUUID());
+        dataSet.setDefinition(def);
+        staticDataSetProvider.registerDataSet(dataSet);
+        return staticDataSetProvider.lookupDataSet(def, lookup);
+    }
+
+    protected DataSet toDataSet(QueryResponse response) {
+        DataSet dataSet = DataSetFactory.newEmptyDataSet();
+        List<Result> results = response.getResults();
+
+        Set<String> metricColumns = getMetricColumns(results);
+
+        metricColumns.forEach(c -> dataSet.addColumn(c, ColumnType.LABEL));
+
+        dataSet.addColumn(TIME_COLUMN, ColumnType.NUMBER);
+        dataSet.addColumn(VALUE_COLUMN, response.getResultType() == ResultType.STRING
+                ? ColumnType.TEXT
+                : ColumnType.NUMBER);
+
+        for (Result result : results) {
+            for (Value value : result.getValues()) {
+                Map<String, String> metric = result.getMetric();
+                Object[] row = new Object[metric.size() + 2];
+                int i = 0;
+                for (String key : metricColumns) {
+                    row[i++] = metric.get(key);
+                }
+                row[i++] = value.getTimestamp();
+                row[i] = value.getValue();
+                dataSet.addValuesAt(dataSet.getRowCount(), row);
+            }
+        }
+        return dataSet;
```

Comment:
I don't like 3 nested `for` loops but I guess there is no other way, right?

Reply:
Unfortunately no, for each metric column we add a dataset column, hence why we have the loop from line 122 - but it is expected to have less than 10 records as metric columns